book-content/chapters/13-modules-scripts-declaration-files.md

```diff
@@ -0,0 +1,988 @@
+
+By installing a package with `@types/*` and your library as a dev dependency, you can add type definitions that TypeScript will be able to use immediately.
+
+For example, say you're using the `diff` library to check for the difference between two strings:
+
+```tsx
+import Diff from "diff"; // red squiggly line under "diff"
+
+const message1 = "Now playing: 'Run Run Run'";
+const message2 = "Now playing: 'Bye Bye Bye'";
+
+const differences = Diff.diffChars(message1, message2);
+```
+
+TypeScript reports an error underneath the `import` statement because it can't find type definitions, even though the library is installed over 40 million times a week from NPM:
+
+```txt
+hovering over "diff" shows:
+Could not find a declaration file for module 'diff'. Try `npm install --save-dev @types/diff` if it exists or add a new declaration (.d.ts) file containing `declare module 'diff';`
+```
+
+Since we're using `pnpm` instead of `npm`, our installation command looks like this:
+
+```bash
+pnpm i -D @types/diff
+```
+
+Once the type definitions from DefinitelyTyped are installed, TypeScript will recognize the `diff` library and provide type checking and autocompletion for it:
+
+```tsx
+// hovering over differences shows:
+const differences: Diff.Change[];
+```
+
+This is a great solution for libraries that haven't been updated in a while, or for more commonly-used libraries (like, say, React) that don't ship with type definitions.
+
+### `skipLibCheck`
+
+As we've seen, your project can contain hundreds of declaration files. By default, TypeScript considers these files as part of your project. So, it checks them for type errors every single time.
+
+This can result in extremely frustrating situations where a type error in a third-party library can prevent your project from compiling.
+
+To avoid this, TypeScript has a `skipLibCheck` setting. When set to `true`, TypeScript will skip checking declaration files for type errors.
+
+```json
+{
+  "compilerOptions": {
+    "skipLibCheck": true
+  }
+}
+```
+
+This is a must-have in any TypeScript project because of the sheer number of declaration files that are included. Adding this setting speeds up compilation and prevents unnecessary errors.
+
+#### The Downsides of `skipLibCheck`
+
+`skipLibCheck` comes with one enormous downside, though. It doesn't just skip declaration files in `node_modules` - it skips _all_ declaration files.
+
+This means that if you make a mistake authoring a declaration file, TypeScript won't catch it. This can lead to bugs that are difficult to track down.
+
+This is one of my main gripes with TypeScript - `skipLibCheck` is a must-have, because of the danger of incorrect third-party declaration files. But it also makes authoring your own declaration files much harder.
```

Comment:
Playing devil's advocate - if your 3rd party declaration file has an error and you suppress it... how do you know that it won't affect u in some under-specified way? You are effectively rolling a die at this point to see when and how it might affect you.